File: exercises/day19-North-Pole-Companies/scala/src/test/scala/com/northpole/InvoicePrinterTest.scala
package com.northpole

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import org.approvaltests.Approvals
import org.scalatest.funsuite.AnyFunSuite

import java.nio.file.{Files, Paths}
import scala.jdk.CollectionConverters._

class InvoicePrinterTest extends AnyFunSuite {

  test("exampleInvoice") {
    val elfCompanies = loadElfCompanies()
    val invoice = loadInvoice()
    val printer = new InvoicePrinter()

    val result = printer.print(invoice, elfCompanies)

    Approvals.verify(result)
  }

  // TODO: Add exampleInvoiceWithTaxes test here

  private def loadElfCompanies(): Map[String, ElfCompany] = {
    val json = Files.readString(Paths.get("src/test/resources/elfCompanies.json"))
    val gson = new Gson()
    val typeToken = new TypeToken[java.util.Map[String, java.util.Map[String, String]]]() {}
    val rawData: java.util.Map[String, java.util.Map[String, String]] = 
      gson.fromJson(json, typeToken.getType)

    rawData.asScala.map { case (key, data) =>
      key -> ElfCompany(
        name = data.get("name"),
        `type` = data.get("type"),
        region = data.get("region")
      )
    }.toMap
  }

  private def loadInvoice(): Invoice = {
    val json = Files.readString(Paths.get("src/test/resources/order.json"))
    val gson = new Gson()
    val typeToken = new TypeToken[java.util.Map[String, Object]]() {}
    val orderData: java.util.Map[String, Object] = gson.fromJson(json, typeToken.getType)

    val customer = orderData.get("customer").asInstanceOf[String]
    val deliveriesData = orderData.get("deliveries")
      .asInstanceOf[java.util.List[java.util.Map[String, Object]]]

    val deliveries = deliveriesData.asScala.map { deliveryData =>
      Delivery(
        companyID = deliveryData.get("companyID").asInstanceOf[String],
        packages = deliveryData.get("packages").asInstanceOf[Double].toInt
      )
    }.toList

    Invoice(customer, deliveries)
  }
}

File: exercises/day19-North-Pole-Companies/typescript/test/InvoicePrinter.test.ts
import { InvoicePrinter } from '../src/InvoicePrinter';
import { Invoice } from '../src/Invoice';
import { ElfCompany } from '../src/ElfCompany';
import * as fs from 'fs';
import { verifyAsJSON } from 'approvals';

describe('InvoicePrinter', () => {
  test('exampleInvoice', () => {
    const elfCompanies = loadElfCompanies();
    const invoice = loadInvoice();
    const printer = new InvoicePrinter();

    const result = printer.print(invoice, elfCompanies);

    verifyAsJSON(__dirname, result);
  });

  // TODO: Add exampleInvoiceWithTaxes test here

  function loadElfCompanies(): Map<string, ElfCompany> {
    const json = fs.readFileSync('test/resources/elfCompanies.json', 'utf8');
    const data = JSON.parse(json);
    const companies = new Map<string, ElfCompany>();

    for (const [key, value] of Object.entries(data)) {
      const company = value as any;
      companies.set(key, {
        name: company.name,
        type: company.type,
        region: company.region
      });
    }
    return companies;
  }

  function loadInvoice(): Invoice {
    const json = fs.readFileSync('test/resources/order.json', 'utf8');
    return JSON.parse(json);
  }
});
